### ISSUE TITLE
HexChat usesthe lowest IP number in its DNS lookups

### ISSUE THREAD

**Issue by danielg4** [hexchat/hexchat#1767]

**Comment by Arnavion** [comment]
Their guess is incorrect. HC does cycle through every server in the DNS response.

**Comment by danielg4** [comment]
Cycle through every server? Why not just do what web browsers do?

**Comment by Arnavion** [comment]
What exactly is it you think web browsers do?

**Comment by danielg4** [comment]
Well, there is no way they reuse the same address every time when DNS resolves to multiple. That would defeat the point of such DNS records.

**Comment by Arnavion** [comment]
The web browser equivalent of your situation is that you go to google.com and have the page load but with a JS error, then expect subsequent page reloads to try a different IP because of that. Web browsers are within their rights to reuse the same IP as long as an HTTP server responded on it (ignoring such things as TTL).

**Comment by danielg4** [comment]
In HTTP 1.1, the connection is still open in such a case. Freenode was explicitly closing it.

**Comment by TingPing** [comment]
As far as hexchat is concerned that connection was fine it doesn't have much of a way to know "authentication failed because this server lost connection to the authentication server, etc...".

**Comment by danielg4** [comment]
So, a connection being fine indicates its IP address is preferred over other equal addresses? No.

**Comment by TingPing** [comment]
Well firstly I am pretty sure hexchat makes no effort to try other IP's from DNS and it's only feature related to this is jumping servers if you have multiple. For the latter the best we could do is try a random one and for the latter yes a connection being fine indications it shouldn't jump IMO.

**Comment by danielg4** [comment]
Having multiple servers specified in the server tab and having multiple servers specified in DNS are **not** the same situation. Not picking a random IP from DNS is itself a bug in any case, regardless of anything else.

**Comment by TingPing** [comment]
I agree they are different features but I'm just saying that is all hexchat has.

> Not picking a random IP from DNS is itself a bug in any case, regardless of anything else.

According to a comment in the source they seem to expect `getaddrinfo()` to do this already:

> Many IRC networks rely on round-robin DNS for load balancing, rotating the list
of IP address on each query. However, this method breaks when DNS queries are
cached. Mac OS X and Darwin handle DNS lookups through the lookupd daemon, which
caches queries in its default configuration: thus, if we always pick the first
address, we will be stuck with the same host (which might be down!) until the
TTL reaches 0 or lookupd is reset (typically, at reboot). Therefore, we need to
pick a random address from the result list, instead of always using the first.

I'm not sure if Windows caches by default now or if this was ever true on it.

**Comment by danielg4** [comment]
I'm on Ubuntu.

**Comment by Arnavion** [comment]
To be honest I don't see any particular reason why we only do it for LOOKUPD. May as well do it for everything. It would solve both the problems mentioned in this thread.

**Comment by danielg4** [comment]
Sounds like a plan!

**Comment by danielg4** [comment]
Just checking in: Was the change that @Arnavion suggested ever made?
